src/Couchbase.Extensions.Encryption/EncryptedFieldConverter.cs

```diff
@@ -34,98 +34,125 @@ public EncryptableFieldConverter(PropertyInfo targetProperty, Dictionary<string,
         public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
         {
             var rawJson = JsonConvert.SerializeObject(value, SerializerSettings);
-            var cryptoProvider = CryptoProviders[ProviderName];
 
-            var rawBytes = Encoding.UTF8.GetBytes(rawJson);
-            var cipherText = cryptoProvider.Encrypt(rawBytes, out var iv);
-            var base64CipherText = Convert.ToBase64String(cipherText);
-
-            string base64Iv = null;
-            if (iv != null)
+            if (!CryptoProviders.TryGetValue(ProviderName, out var cryptoProvider))
             {
-                base64Iv = Convert.ToBase64String(iv);
+                throw new CryptoProviderNotFoundException(ProviderName);
             }
 
-            byte[] signatureBytes = null;
-            if (cryptoProvider.RequiresAuthentication)
+            var rawBytes = Encoding.UTF8.GetBytes(rawJson);
+            try
             {
-                //sig = HMAC256(BASE64(kid + alg + iv + ciphertext))
-                var kidBytes = Encoding.UTF8.GetBytes(cryptoProvider.PublicKeyName);
-                var algBytes = Encoding.UTF8.GetBytes(cryptoProvider.AlgorithmName);
-                var buffer = new byte[kidBytes.Length + algBytes.Length + iv.Length + cipherText.Length];
-
-                Buffer.BlockCopy(kidBytes, 0, buffer, 0, kidBytes.Length);
-                Buffer.BlockCopy(algBytes, 0, buffer, kidBytes.Length, algBytes.Length);
-                Buffer.BlockCopy(iv, 0, buffer, kidBytes.Length + algBytes.Length, iv.Length);
-                Buffer.BlockCopy(cipherText, 0, buffer, kidBytes.Length + algBytes.Length + iv.Length, cipherText.Length);
-
-                //sign the entire buffer
-                signatureBytes = cryptoProvider.GetSignature(buffer);
-            }
+                var cipherText = cryptoProvider.Encrypt(rawBytes, out var iv);
 
-            var token = new JObject(
-                new JProperty("alg", cryptoProvider.AlgorithmName),
-                new JProperty("kid", cryptoProvider.PublicKeyName),
-                new JProperty("ciphertext", base64CipherText));
+                var base64CipherText = Convert.ToBase64String(cipherText);
 
-            if (signatureBytes != null)
-            {
-                var base64Sig = Convert.ToBase64String(signatureBytes);
-                token.Add("sig", base64Sig);
+                string base64Iv = null;
```

Comment:
This assignment could be moved lower to only assign if iv is not null - same signature bytes